test/common/http/conn_manager_impl_test.cc

```diff
@@ -2835,6 +2835,136 @@ TEST_F(HttpConnectionManagerImplTest, RequestTimeoutIsDisarmedOnConnectionTermin
   filter_callbacks_.connection_.raiseEvent(Network::ConnectionEvent::RemoteClose);
 }
 
+TEST_F(HttpConnectionManagerImplTest, RequestHeaderTimeoutValidlyConfigured) {
+  request_headers_timeout_ = std::chrono::milliseconds(10);
+  setup(false, "");
+
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance&) -> Http::Status {
+    Event::MockTimer* request_header_timer = setUpTimer();
+    EXPECT_CALL(*request_header_timer, enableTimer(request_headers_timeout_, _));
+    EXPECT_CALL(*request_header_timer, disableTimer());
+
+    conn_manager_->newStream(response_encoder_);
+    return Http::okStatus();
+  }));
+
+  Buffer::OwnedImpl fake_input("1234");
+  conn_manager_->onData(fake_input, false);
+
+  expectOnDestroy();
+  filter_callbacks_.connection_.raiseEvent(Network::ConnectionEvent::RemoteClose);
+}
+
+TEST_F(HttpConnectionManagerImplTest, RequestHeaderTimeoutCallbackDisarmsAndReturns408) {
+  request_headers_timeout_ = std::chrono::milliseconds(10);
+  setup(false, "");
+
+  std::string response_body;
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance&) -> Http::Status {
+    Event::MockTimer* request_header_timer = setUpTimer();
+    EXPECT_CALL(*request_header_timer, enableTimer(request_headers_timeout_, _)).Times(1);
+    EXPECT_CALL(*request_header_timer, disableTimer()).Times(AtLeast(1));
+
+    EXPECT_CALL(response_encoder_, encodeHeaders(_, false))
+        .WillOnce(Invoke([](const ResponseHeaderMap& headers, bool) -> void {
+          EXPECT_EQ("408", headers.getStatusValue());
+        }));
+    EXPECT_CALL(response_encoder_, encodeData(_, true)).WillOnce(AddBufferToString(&response_body));
+
+    conn_manager_->newStream(response_encoder_);
+    EXPECT_CALL(filter_callbacks_.connection_.dispatcher_, setTrackedObject(_)).Times(2);
+    request_header_timer->invokeCallback();
+    return Http::okStatus();
+  }));
+
+  Buffer::OwnedImpl fake_input("1234");
+  conn_manager_->onData(fake_input, false); // kick off request
+
+  EXPECT_EQ(1U, stats_.named_.downstream_rq_header_timeout_.value());
+  EXPECT_EQ("request header timeout", response_body);
+}
+
+TEST_F(HttpConnectionManagerImplTest,
+       RequestHeaderTimeoutIsNotDisarmedOnIncompleteRequestWithHeader) {
+  request_headers_timeout_ = std::chrono::milliseconds(10);
+  setup(false, "");
+
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance&) -> Http::Status {
+    Event::MockTimer* request_header_timer = setUpTimer();
+    EXPECT_CALL(*request_header_timer, enableTimer(request_headers_timeout_, _)).Times(1);
+    EXPECT_CALL(*request_header_timer, disableTimer()).Times(1);
+
+    RequestDecoder* decoder = &conn_manager_->newStream(response_encoder_);
+    RequestHeaderMapPtr headers{
+        new TestRequestHeaderMapImpl{{":authority", "host"}, {":path", "/"}, {":method", "GET"}}};
+
+    // the second parameter 'false' leaves the stream open
+    decoder->decodeHeaders(std::move(headers), false);
```

Comment:
There's just too much mocking happening on these "tests".  I don't follow what you're testing here or how it relates to the call to conn_manager_->onData below.

Reply:
That's reasonable, because the tests were bad. PTAL at the cleaned-up version.